pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,835 @@
+
+        self._measurements = []
+        """list[~.MeasurementProcess]: measurement processes recorded by the tape."""
+
+        self._par_info = {}
+        """dict[int, dict[str, Operation or int]]: Parameter information. Keys are
+        parameter indices (in the order they appear on the tape), and values are a
+        dictionary containing the corresponding operation and operation parameter index."""
+
+        self._trainable_params = set()
+        self._graph = None
+        self._output_dim = 0
+
+        self.wires = qml.wires.Wires([])
+        self.num_wires = 0
+
+        self.jacobian_options = {}
+
+        self.hash = 0
+        self.is_sampled = False
+
+        self._stack = None
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__}: wires={self.wires.tolist()}, params={self.num_params}>"
+
+    def __enter__(self):
+        if not QueuingContext.recording():
+            # if the tape is the first active queuing context
+            # monkeypatch the operations to support the new queuing context
+            with contextlib.ExitStack() as stack:
+                for mock in mock_operations():
+                    stack.enter_context(mock)
+                self._stack = stack.pop_all()
+
+        QueuingContext.append(self)
+        return super().__enter__()
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        super().__exit__(exception_type, exception_value, traceback)
+
+        if not QueuingContext.recording():
+            # remove the monkeypatching
+            self._stack.__exit__(exception_type, exception_value, traceback)
+
+        self._process_queue()
+
+    @property
+    def interface(self):
+        """str, None: automatic differentiation interface used by the quantum tape (if any)"""
+        return None
+
+    # ========================================================
+    # construction methods
+    # ========================================================
+
+    def _process_queue(self):
+        """Process the annotated queue, creating a list of quantum
+        operations and measurement processes.
+
+        This method sets the following attributes:
+
+        * ``_ops``
+        * ``_measurements``
+        * ``_par_info``
+        * ``_output_dim``
+        * ``_trainable_params``
+        * ``is_sampled``
+        """
+        self._prep = []
+        self._ops = []
+        self._measurements = []
+        self._output_dim = 0
+
+        for obj, info in self._queue.items():
+
+            if isinstance(obj, QuantumTape):
+                self._ops.append(obj)
+
+            elif isinstance(obj, qml.operation.Operation) and not info.get("owner", False):
+                # operation objects with no owners
+
+                if self._measurements:
+                    raise ValueError(
+                        f"Quantum operation {obj} must occur prior to any measurements."
+                    )
```

Comment:
When could this happen?

Reply:
e.g.,

```python
with QuantumTape() as tape:
    expval(qml.PauliZ(0))
    qml.RX(0.56, wires=1)
```

It essentially forces all operations to precede all measurements in the queue (even if an operation and measurement are 'independent' as above). This is the same restriction we have currently with the QNode, but we never really noticed it since the `return measurements` pattern was so entrenched.

Reply:
Hi
I try use:
@qml.qnode(dev)
def circuit1(N: int):
  qml.PauliX(0)
  w = range(0, 1)
  result = qml.expval(qml.PauliX(0))
  qml.Hadamard(wires=[1])
  return result

but got:
ValueError: Quantum operation Hadamard(wires=[1]) must occur prior to any measurements.

qml.version() = 0.20.0

It it an old issue or it need create a new one?
Thanks
@flatslm

Reply:
Hi @flatslm! This occurs because `qml.expval()` must always be the last instruction in the QNode. For example, the following should work:

```python
@qml.qnode(dev)
def circuit1(N: int):
    qml.PauliX(0)
    qml.Hadamard(wires=[1])
    return qml.expval(qml.PauliX(0))
```

For future reference, I suggest posting questions on the [PennyLane discussion forum ](https://discuss.pennylane.ai/) rather than GitHub PRs, as they will be much more visible 🙂